src/facet/inspection/_shap_projection.py

```diff
@@ -0,0 +1,379 @@
+"""
+Projection of SHAP contribution scores (i.e, SHAP importance) of all possible
+pairings of features onto the SHAP importance vector in partitions of for synergy,
+redundancy, and independence.
+"""
+import logging
+from typing import List, Optional, TypeVar
+
+import numpy as np
+import pandas as pd
+
+from pytools.api import AllTracker, inheritdoc
+
+from ._shap import ShapCalculator, ShapInteractionValuesCalculator
+from ._shap_global_explanation import (
+    AffinityMatrices,
+    ShapGlobalExplainer,
+    ShapInteractionGlobalExplainer,
+    cov,
+    cov_broadcast,
+    diagonal,
+    ensure_last_axis_is_fast,
+    fill_diagonal,
+    transpose,
+)
+
+log = logging.getLogger(__name__)
+
+__all__ = [
+    "ShapInteractionProjector",
+    "ShapProjector",
+]
+
+#: if ``True``, optimize numpy arrays to ensure pairwise partial summation.
+#: But given that we will add floats of the same order of magnitude and only up
+#: to a few thousand of them in the base case, the loss of accuracy with regular
+#: (sequential) summation will be negligible in practice
+_PAIRWISE_PARTIAL_SUMMATION = False
+
+#
+# Type variables
+#
+
+T_Self = TypeVar("T_Self")
+
+#
+# Ensure all symbols introduced below are included in __all__
+#
+
+__tracker = AllTracker(globals())
+
+
+#
+# Class definitions
+#
+
+
+@inheritdoc(match="""[see superclass]""")
+class ShapProjector(ShapGlobalExplainer):
+    """
+    Decomposes SHAP vectors (i.e., SHAP contribution) of all possible pairings
+    of features into additive components for association and independence.
+    SHAP contribution scores are calculated as the standard deviation of the individual
+    interactions per observation. Using this metric, rather than the mean of absolute
+    interactions, allows us to calculate the decomposition without ever constructing
+    the decompositions of the actual SHAP vectors across observations.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.association_: Optional[AffinityMatrices] = None
+
+    def association(self, absolute: bool, symmetrical: bool) -> np.ndarray:
+        """[see superclass]"""
+        self._ensure_fitted()
+        return self.association_.get_matrix(symmetrical=symmetrical, absolute=absolute)
+
+    def _fit(self, shap_calculator: ShapCalculator) -> None:
+        #
+        # basic definitions
+        #
+
+        shap_values: pd.DataFrame = shap_calculator.get_shap_values(consolidate="mean")
+        n_outputs: int = len(shap_calculator.output_names_)
```

Comment:
```suggestion
        # name of target columns (only one if binary classifier)
        n_outputs: int = len(shap_calculator.output_names_)
```

Reply:
You mean, number of target columns? (not name)

Reply:
yes :) 
I meant to specify the shap_calculator.output_names_ refers to the list of target names

Reply:
Ah - not always which is why I call it outputs. This is more general than targets:

- In multi-class classification, there is one output column for every class although all of these relate to the same target.
- In multi-target regression, the outputs are indeed the targets.